### ISSUE TITLE
Disallow super() and .this in static methods

### ISSUE THREAD

**Issue by Torhal** [sindresorhus/eslint-plugin-unicorn#1050]
The `this` keyword on static methods refers the class instance (the constructor). This can be confusing.

## Fail

```js
class A {
    static foo() {
        doSomething()
    }

    static bar() {
        this.foo()   //ERROR: Unexpected 'this'.
    }
}

class B extends A {
    static foo() {
        super.foo()  //ERROR: Unexpected 'super'.
    }
}
```

## Pass

```js
class A {
    static foo() {
        doSomething()
    }

    static bar() {
        A.foo()
    }
}

class B extends A {
    static foo() {
        A.foo()
    }
}
```

**Comment by sindresorhus** [comment]
Why is `this` confusing on a static method?

**Comment by Torhal** [comment]
Well, it _can_ be confusing to people switching between languages or are only used to languages where using ```this``` in a static context isn't allowed because it's reserved for class instances where here it's referring to the constructor.

It can also become imprecise when used with extended classes when a static ```this``` of a parent class no longer specifically refers to the parent class.

**Comment by sindresorhus** [comment]
I would actually want the opposite, so I think the rule should let you enforce either `this` or "class name" in static methods. Same with `super`.

Any suggestions for the rule name?

**Comment by fisker** [comment]
Without "classname" how to access another property/method?

**Comment by sindresorhus** [comment]
> Without "classname" how to access another property/method?

Not sure I understand the question. I would access it with `this` or `super`.

**Comment by fisker** [comment]
Oh, sorry I misunderstood, I though this rule not allow `this` "class name" and `super`.

**Comment by blujedis** [comment]
Something to consider.

When working in Typescript there are cases where you need to create or wrap things in anonymous classes. This is because you're passing down the generic type to the underlying static methods. More complex scenarios.

You may not have access to a class name. Not sure there's a likely default here. 

With @sindresorhus on this as to enforcing either this or classname but this particular rule I can see being set per file in several cases.

**Comment by sindresorhus** [comment]
Suggestions for the rule name:

- `this-and-super-in-static-methods`
- `class-reference-in-static-methods`

Better suggestions welcome.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by Torhal** [comment]
> * `class-reference-in-static-methods`

That seems to be the best and most concise for what this has turned into. The original proposal would probably have been something like `no-static-this`.

**Comment by fisker** [comment]
👍  for `class-reference-in-static-methods`

**Comment by sindresorhus** [comment]
If anyone wants to work on this, see the initial attempt and feedback in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1119.
